src/dos/dos_memory.cpp

```diff
@@ -18,23 +18,90 @@
 
 #include "dosbox.h"
 
-#include "mem.h"
 #include "dos_inc.h"
+#include "enum.h"
+#include "mem.h"
 
-#define UMB_START_SEG 0x9fff
+enum class McbFaultStrategy { deny, repair, report, allow };
 
-static uint16_t memAllocStrategy = 0x00;
+// Constants
+// ~~~~~~~~~
 
-static void DOS_CompressMemory(void) {
-	uint16_t mcb_segment=dos.firstMCB;
+// MCB - DOS Memory Control Block Format
+// Offset Size   Description
+// 00     byte   'M' 4Dh  member of a MCB chain, (not last)
+//               'Z' 5Ah  indicates last entry in MCB chain
+// - the 'M' and 'Z' are said to represent Mark Zbikowski
+constexpr uint8_t mcb_type_M = {0x4d};
+constexpr uint8_t mcb_type_Z = {0x5a};
+
+// Upper member block starting segment
+constexpr uint16_t UMB_START_SEG = {0x9fff};
+
+static uint16_t memAllocStrategy = {0x00};
+
+constexpr uint16_t max_allowed_faults = {100};
+// not based on anything in particular. Player Manager 2 requires ~17 corrections.
+
+static auto mcb_fault_strategy = McbFaultStrategy::repair;
+
+void DOS_SetMcbFaultStrategy(const char *mcb_fault_strategy_pref)
+{
+	assert(mcb_fault_strategy_pref);
+	mcb_fault_strategy =
+	        enum_cast<McbFaultStrategy>(mcb_fault_strategy_pref).value();
+}
+
+// returns true if the MCB block needed triaging
+static bool triage_block(DOS_MCB &mcb, const uint8_t repair_type)
+{
+	auto mcb_type_is_valid = [&]() -> bool {
+		const auto t = mcb.GetType();
+		return t == mcb_type_M || t == mcb_type_Z;
+	};
+	if (mcb_type_is_valid())
+		return false;
+
+	switch (mcb_fault_strategy) {
+	case McbFaultStrategy::deny:
+		E_Exit("DOS_MEMORY: Exiting due corrupt MCB chain");
+		break;
+
+	case McbFaultStrategy::repair:
+		LOG_INFO("DOS_MEMORY: Repairing MCB block in segment %04xh from type '%02x' to '%02x'",
+		         mcb.GetPSPSeg(),
+		         mcb.GetType(),
+		         repair_type);
+		mcb.SetType(repair_type);
+		assert(mcb_type_is_valid());
+		break;
+
+	case McbFaultStrategy::report:
+		LOG_WARNING("DOS_MEMORY: Reporting MCB block in segment %04xh with corrupt type '%02x'",
+		            mcb.GetPSPSeg(),
+		            mcb.GetType());
+		break;
+
+	case McbFaultStrategy::allow:
+	default: break; // do nothing
+	}
+	return true;
+}
+
+static void DOS_CompressMemory()
+{
+	uint16_t mcb_segment = dos.firstMCB;
 	DOS_MCB mcb(mcb_segment);
 	DOS_MCB mcb_next(0);
 
-	while (mcb.GetType()!=0x5a) {
-		mcb_next.SetPt((uint16_t)(mcb_segment+mcb.GetSize()+1));
-		if (GCC_UNLIKELY((mcb_next.GetType()!=0x4d) && (mcb_next.GetType()!=0x5a))) E_Exit("Corrupt MCB chain");
-		if ((mcb.GetPSPSeg()==MCB_FREE) && (mcb_next.GetPSPSeg()==MCB_FREE)) {
-			mcb.SetSize(mcb.GetSize()+mcb_next.GetSize()+1);
+	uint16_t faults = {0u};
```

Comment:
What's the benefit of the curly braces around simple value assignments?

Reply:
It gives a compiler error if the RHS is being narrowed or losing data (and yeah, nothing lost here!).

But it's pretty handy:

``` c++
    int a = 11.2;
    int b = {11.2};
```

First is permitted, but second gives this error:

``` text
main.cpp:16:18: error: narrowing conversion of ‘1.1199999999999999e+1’ from ‘double’ to ‘int’ [-Wnarrowing]
   16 |     int b = {11.2};
```


Becomes very useful when the RHS is a template T or `auto` set by some other RHS result.

Reply:
I see. It's a pity you need to use curly braces for that, that adds a lot of noise. I wonder if it's possible to turn this on as an optional warning without having to litter the codebase with curly braces everywhere.

Reply:
its possible but then you get +3000 warnings from the dosbox code :)

Reply:
Indeed @johnnovak ; @LowLevelMahn investigated this narrowing, and I tried coming at it from a warning perspective (instead of using the language-spec itself), under this PR: https://github.com/dosbox-staging/dosbox-staging/pull/1704

> Its possible but then you get +3000 warnings from the dosbox code :)

We're making progress using the per-file approach (`checks.h`), and these sources are guaranteed to be type-narrowing correct:

**`git grep CHECK_NARROWING`**

```
src/hardware/adlib_gold.cpp:CHECK_NARROWING();
src/hardware/innovation.cpp:CHECK_NARROWING();
src/hardware/pcspeaker_discrete.cpp:CHECK_NARROWING();
src/hardware/pcspeaker_impulse.cpp:CHECK_NARROWING();
src/hardware/serialport/serialmouse.cpp:CHECK_NARROWING();
src/ints/mouse.cpp:CHECK_NARROWING();
src/ints/mouse_dos_driver.cpp:CHECK_NARROWING();
src/ints/mouse_ps2_bios.cpp:CHECK_NARROWING();
src/ints/mouse_serial.cpp:CHECK_NARROWING();
src/ints/mouse_vmware.cpp:CHECK_NARROWING();
src/libs/zmbv/zmbv.cpp:CHECK_NARROWING();
```

@johnnovak , I think you added that one to Gold :+1: , and @FeralChild64 's overhauled the mouse land entirely :+1: !

Reply:
When the all of the sources are compliant we can turn on those flags at the project-level and strip away the per-file checks.